runtime/gc_modron_startup/mminit.cpp

```diff
@@ -3018,6 +3018,38 @@ gcInitializeDefaults(J9JavaVM* vm)
 		if ((-1 != memoryParameterTable[opt_Xms]) && (extensions->initialMemorySize > extensions->memoryMax)) {
 			extensions->memoryMax = extensions->initialMemorySize;
 		}
+	} else {
+#if defined(J9ZOS390) && !defined(OMR_ENV_DATA64)
+		/*
+		 *  Maximum heap size for ZOS 31-bit should not exceed 2047m
+		 *  This explicit check is necessary to prevent ZOS failure at
+		 *  attempt to allocate too large piece of memory.
+		 *  Please note that size of memory can be allocated in reality
+		 *  is much smaller, something about 1750m. An allocation request
+		 *  in the higher range under 2047m is going to fail normally.
+		 *  The error message here is exactly the same as in regular failure.
+		 */
+#define MAX_HEAP_SIZE_FOR_ZOS31	((UDATA)2047 * 1024 * 1024)
+		if (MAX_HEAP_SIZE_FOR_ZOS31 < extensions->memoryMax) {
+			/* Obtain the qualified size */
+			UDATA size = extensions->memoryMax;
+			const char* qualifier = NULL;
+			qualifiedSize(&size, &qualifier);
+
+			const char *format = j9nls_lookup_message(
+				J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
+				J9NLS_GC_FAILED_TO_INSTANTIATE_HEAP_SIZE_REQUESTED,
+				"Failed to instantiate heap; %zu%s requested");
```

Comment:
perhaps we could be more specific like 'Max ZOS31 process memory limit exceeded;....' if that makes sense

Reply:
My intention was to make output undistinguishable between cases failure attempt to allocate below 2g and above 2g bars. And I am trying to avoid to add another NLS message. If you think we need to do it I can change.

Reply:
Just a suggestion. Ultimately up to you, since you are more familiar with what sorts of error messages we already generate and how detailed we are.